assembl/static2/js/app/components/debate/common/ideasLevel.jsx

```diff
@@ -0,0 +1,408 @@
+        }
+      }
+    }
+  }
+
+  scrollToIdeaIndexIfNecessary(index) {
+    const thematicWidth = this.getOption('thematicWidth');
+    const carouselWidth = this.getOption('carouselWidth');
+    const ideaInitialX = index * thematicWidth;
+    if (ideaInitialX + thematicWidth > carouselWidth) {
+      const numberOfIdeasFullyShown = Math.floor(carouselWidth / thematicWidth);
+      const middleIdeaX = Math.floor(numberOfIdeasFullyShown / 2) * thematicWidth;
+      const targetScrollValue = -1 * (index * thematicWidth - middleIdeaX);
+      this.moveAllChildren(this.getRow(), targetScrollValue, true);
+    }
+  }
+
+  updateScrollButtonsVisibility(isInline, isAnimatingTowardsInline) {
+    const thematicWidth = this.getOption('thematicWidth');
+    const parentWidth = this.getOption('parentWidth');
+    const numberOfThematics = this.getOption('numberOfThematics');
+    let targetScrollLeftVisibility = false;
+    let targetScrollRightVisibility = false;
+
+    if (!(isInline || isAnimatingTowardsInline)) {
+      // if ideas do not display inline, then do not show any scroll icon
+      targetScrollLeftVisibility = false;
+      targetScrollRightVisibility = false;
+    } else if (numberOfThematics * thematicWidth < parentWidth) {
+      // if ideas show inline and all ideas can show in the carousel with no need to scroll, then do not show any scroll icon
+      targetScrollLeftVisibility = false;
+      targetScrollRightVisibility = false;
+    } else {
+      const targetValueInt = this.getOption('scrollDisplacement');
+      const displacementMin = this.getOption('displacementMin');
+      const displacementMax = this.getOption('displacementMax');
+      if (targetValueInt >= displacementMax) {
+        targetScrollLeftVisibility = false;
+      } else {
+        targetScrollLeftVisibility = true;
+      }
+
+      if (targetValueInt <= displacementMin) {
+        targetScrollRightVisibility = false;
+      } else {
+        targetScrollRightVisibility = true;
+      }
+    }
+
+    /*
+    FIXME: Using this.setState() with isScrollLeftButtonVisible and
+    isScrollRightButtonVisible does not show correct buttons UI, why?
+    Below is a workaround.
+    */
+    if (!this.scrollLeft || !this.scrollRight) {
+      const f = () => {
+        this.updateScrollButtonsVisibility(isInline, isAnimatingTowardsInline);
+      };
+      f.bind(this);
+      setTimeout(f, 500);
+    } else {
+      this.scrollLeft.setState({ isVisible: targetScrollLeftVisibility });
+      this.scrollLeft.forceUpdate(); // FIXME: Why is this needed? component should re-render and show correct UI by itself
+      this.scrollRight.setState({ isVisible: targetScrollRightVisibility });
+      this.scrollRight.forceUpdate(); // FIXME: Why is this needed? component should re-render and show correct UI by itself
+    }
+  }
+
+  initializeConstants() {
+    let parentWidth = 1400;
+    if (this.me && this.me.parentNode && this.me.parentNode.clientWidth) {
```

Comment:
It would not be ideal either: I think the parent component does not know its clientWidth at the moment when execution is in its its render() method and it's calling the child component.